### PR TITLE
Add refined tupled records prototype

### PR DESCRIPTION
**Pull request by biboudis**
This PR adds two macros, `toTuple` and  `fromTuple` that transform a refinement type into a tuple and vice-versa:
 
```scala
  type Person = Record {
    val name: String
    val age: Int
  }

  val person: Person = Record("name" -> "Emma", "age" -> 42).asInstanceOf[Person]
  val res: (("name", String), ("age", Int)) = person.toTuple
  val res2: Person = Record.fromTuple(res)
```

To enable this functionality the user would need to `SelectableRecord` instead of `Selectable`, e.g., :

```scala
  case class Record(elems: (String, Any)*) extends SelectableRecord {
    def selectDynamic(name: String): Any = elems.find(_._1 == name).get._2
  }

  object Record extends SelectableRecordCompanion[Record] {
    def fromUntypedTuple(elems: (String, Any)*): Record = Record(elems: _*)
  }
```

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 1]
```diff
@@ -1258,6 +1265,13 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
       case _ => None
   }
 
+  def Refinement_apply(parent: Type, name: String, info: TypeOrBounds /* Type | TypeBounds */)(given ctx: Context): Refinement = {
+    // fixme: support type refinements (type vs term)
+    //    examine info maybe (TypeOrBounds are used for refinements)
+    //    check aliasing of refinement
+    Types.RefinedType(parent, name.toTermName, info)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]:
```suggestion
     val name1 =
       info match
         case _: TypeBounds => name.toTypeName
         case _ => name.toTermName
    Types.RefinedType(parent, name1, info)
```

File: tests/run-macros/refined-selectable-macro/Test_2.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    println(Record.fromTuple(xxl))
+    // println(Record.fromTuple(("field1" -> 1,"field2" -> 2,"field3" -> 3,"field4" -> 4,"field5" -> 5,"field6" -> 6,"field7" -> 7,"field8" -> 8,"field9" -> 9,"field10" -> 10,"field11" -> 11,"field12" -> 12,"field13" -> 13,"field14" -> 14,"field15" -> 15,"field16" -> 16,"field17" -> 17,"field18" -> 18,"field19" -> 19,"field20" -> 20,"field21" -> 21,"field22" -> 22,"field23" -> 23,"field24" -> 24,"field25" -> 25)))
+
+    println(res2)
+
+    // Record.fromTuple[(("name", String), ("name", Int))]("name" -> "aa", "name" -> 3) // error
```
Comment (nicolasstucki) [inline tests/run-macros/refined-selectable-macro/Test_2.scala]: Still need to extract neg tests

Comment (biboudis) [reply tests/run-macros/refined-selectable-macro/Test_2.scala]: Yeap. Its coming!

**Comment by milessabin** [comment]
I'm not super-keen on the aesthetics of the `asInstanceOf[Person]` in,

```scala
val person: Person = Record("name" -> "Emma", "age" -> 42).asInstanceOf[Person]
```

I appreciate that we need to mention `Person` somewhere, but can't we drop the `asInstanceOf`?

Could,

```scala
val person: Person = Record("name" -> "Emma", "age" -> 42).as[Person]
```
be made to work?

**Comment by nicolasstucki** [comment]
Probably a record should have a better constructor. One that could figure out the refined type.

**Comment by biboudis** [comment]
@nicolasstucki do you mean a smarter implementation of the constructor or something like `Record[Person]`. In that case I think the `as` reveals the intent nicely.

**Comment by nicolasstucki** [comment]
I meant ` Record("name" -> "Emma", "age" -> 42)` would return `Record{val name: String; val age: Int` which is a person

**Comment by biboudis** [comment]
After a quick brainstorming with @liufengyun we propose some functionality on tasty-reflect like the following where `qctx.expected` is something that the macro system provides. WDYT?

```scala
  object Record extends SelectableRecordCompanion[Record] {
    def fromUntypedTuple(elems: (String, Any)*): Record = Record(elems: _*)

    inline def apply(elems: (String, Any)*) <: Record = ${ applyImpl('elem)) }

    def applyImpl(elems: Expr[Seq[(String, Any)]])(given qctx: QuoteContext) = {
      type TT

      implicit val ttype: quoted.Type[TT] = qctx.expected.seal.asInstanceOf[quoted.Type[TT]]
      
      '{ new Record(elems).asInstanceOf[TT] }
    }
  }
```

😃 😃

**Comment by nicolasstucki** [comment]
That is horribly unsound

**Comment by nicolasstucki** [comment]
You probably want something in the lines of 
```scala
inline def apply[R <: Record](elems: (String, Any)*) <: R = ${ applyImpl('elem, '[R])) }
```

**Comment by biboudis** [comment]
I assumed we wanted to avoid the type parameter 😛! `apply[R]` sounds good!

**Comment by julienrf** [comment]
> I appreciate that we need to mention Person somewhere, but can't we drop the asInstanceOf?

Why can’t the following work?

```scala
val person: Person = Record("name" -> "Emma", "age" -> 42)
```

**Comment by biboudis** [comment]
Hmm the compiler cannot infer this at this moment ATM 🤷‍♂  I guess it would need some tweaking in the Typer.

**Comment by nicolasstucki** [comment]
No, the apply method would just be another macro

**Comment by nicolasstucki** [comment]
We are getting off-topic. This PR is about prototyping the core macro logic for refinement types. Any API design concerns are for later.

**Comment by biboudis** [comment]
After the last commit I have two observations to share:

1. @anatoliykmetyuk `typeCheckErrors` seems to include the same error many times on the returned list. I don't know if this is by design because in the check file we have the same error twice as well [typeCheckErrors.check](https://github.com/lampepfl/dotty/blob/master/tests/run-bootstrapped/typeCheckErrors.check). WDYT?
2. @nicolasstucki the inlined apply works when I pass `Person` explicitly as the type argument of the constructor call `Record[Person]("name" -> "Emma", "age" -> 42)`

If I don't, the compiler infers `Nothing` (which later splices in the `asInstanceOf`). This may be an inference problem instead of a macro problem.

**Comment by anatoliykmetyuk** [comment]
Yes, typeCheckErrors may have the same error many times. That's because it takes raw report from the reporters of the parser and typer phases and that's what the reporters have. I don't know why precisely they may have multiple errors so I didn't do any filtering.

**Comment by anatoliykmetyuk** [comment]
The only filtering done in `typeCheckErrors` is to exclude parser errors from typer errors:

https://github.com/lampepfl/dotty/blob/d3a2d61d531aa88532bd9d37392f960c5477b705/compiler/src/dotty/tools/dotc/typer/Inliner.scala#L222

**Comment by biboudis** [comment]
> If I don't, the compiler infers `Nothing` (which later splices in the `asInstanceOf`). This may be an inference problem instead of a macro problem.

Opened #7798

**Pull request merged**